lumicks/pylake/kymo.py

```diff
@@ -2,31 +2,72 @@
 import numpy as np
 
 from .detail.mixin import PhotonCounts
-from .detail.image import reconstruct_image, save_tiff, ImageMetadata
+from .detail.image import reconstruct_image, save_tiff, ImageMetadata, line_timestamps_image
+from .detail.timeindex import to_timestamp
 
 
 class Kymo(PhotonCounts):
     """A Kymograph exported from Bluelake
 
     Parameters
     ----------
-    h5py_dset : h5py.Dataset
-        The original HDF5 dataset containing kymo information
+    name : str
+        Kymograph name
     file : lumicks.pylake.File
-        The parent file. Used to loop up channel data
+        Parent file. Contains the channel data.
+    start : int
+        Start point in the relevant info wave.
+    stop : int
+        End point in the relevant info wave.
+    json : dict
+        Dictionary containing kymograph-specific metadata.
     """
-    def __init__(self, h5py_dset, file):
-        self.start = h5py_dset.attrs["Start time (ns)"]
-        self.stop = h5py_dset.attrs["Stop time (ns)"]
-        self.name = h5py_dset.name.split("/")[-1]
-        self.json = json.loads(h5py_dset[()])["value0"]
+    def __init__(self, name, file, start, stop, json):
+        self.start = start
+        self.stop = stop
+        self.name = name
+        self.json = json
         self.file = file
         self._cache = {}
 
     def __repr__(self):
         name = self.__class__.__name__
         return f"{name}(pixels={self.pixels_per_line})"
 
+    def __getitem__(self, item):
+        """All indexing is in timestamp units (ns)"""
+        if not isinstance(item, slice):
+            raise IndexError("Scalar indexing is not supported, only slicing")
+        if item.step is not None:
+            raise IndexError("Slice steps are not supported")
+
+        start, stop = item.start, item.stop
+        if start is None:
+            start = self.start
+        if stop is None:
+            stop = self.stop
+        start, stop = (to_timestamp(v, self.start, self.stop) for v in (start, stop))
+
+        timestamps = self.infowave.timestamps
+        line_timestamps = line_timestamps_image(timestamps, self.infowave.data, self.pixels_per_line)
+        line_timestamps = np.append(line_timestamps, timestamps[-1])
+
+        i_min = np.searchsorted(line_timestamps, start, side='left')
+        i_max = np.searchsorted(line_timestamps, stop, side='left')
+
+        if i_min >= len(line_timestamps):
+            raise IndexError(f"Slice selection out of range ({(start-self.start)/1e9} sec >= {(self.stop-self.start)/1e9} sec)")
+
+        if i_min >= i_max:
+            raise IndexError(f"Slice would result in empty slice. Valid slice range is from {(start-self.start)/1e9} sec to {(self.stop-self.start)/1e9} sec")
```

Comment:
Is it customary in python to throw in such cases? It's going to be empty but it's not an error state. Genuinely curious!

Reply:
Nope, it's not.

I have changed it to returning an empty kymograph now.